feat(router): Add feature to support the View Transitions API

atscott opened the pull request:
The View Transitions API enables easy animations when transitioning between different DOM states. This commit adds an opt-in feature to the Router which runs the component activation and deactivation logic in the `document.startViewTransition` callback. If the browser does not support this API, route activation and deactivation will happen synchronously.

resolves #49401

atscott commented:
Updating risk to medium since the view transition adds a delay to where code was synchronous before. Exact timings in the Router have historically been fragile. It is technically possible to keep the code synchronous if the feature is not enabled, but avoiding "sometimes synchronous, sometimes asynchronous" branch in the code would be ideal.

ajafff commented on packages/router/src/operators/activate_routes.ts:
```diff
@@ -6,43 +6,66 @@
-          .activate(rootContexts);
-      return t;
+     injector: Injector): MonoTypeOperatorFunction<NavigationTransition> => switchMap(t => {
+      return defer(
```
`defer` should not be necessary, because the callback of `switchMap` is only executed when there is a subscriber.
if you only need it to convert the Promise to Observable, prefer `from` or keep using Promises `switchMap(t => new ActivateRoutes(...).activate(rootContexts).then(() => t))`

I have written similar code multiple times until I realised there is an operator for delaying an emission: `delayWhen(() => new ActivateRoutes(...).activate(rootContexts))`

ajafff commented on packages/router/src/operators/activate_routes.ts:
```diff
@@ -6,43 +6,66 @@
+                           routeReuseStrategy, t.targetRouterState!, t.currentRouterState,
+                           forwardEvent, injector)
+                           .activate(rootContexts))
+          .pipe(mapTo(t));
```
`mapTo` is deprecated in recent versions of rxjs. it's recommended to replace it with `map(() => t)`

ajafff commented on packages/router/src/operators/activate_routes.ts:
```diff
@@ -6,43 +6,66 @@
+      // transition and will do things like emit the `NavigationEnd` event before the View
+      // Transitions happen (resulting in ActivationStart/End events coming after `NavigationEnd`,
+      // for example).
+      return this.transitionHelper(activation.bind(this), this.injector).updateCallbackDone;
```
why do you need to use `.bind(this)` here, but can call it like a plain function 2 lines below?

ajafff commented on packages/router/src/router_module.ts:
```diff
@@ -125,7 +125,7 @@ export class RouterModule {
         config?.preloadingStrategy ? withPreloading(config.preloadingStrategy).ɵproviders : [],
         {provide: NgProbeToken, multi: true, useFactory: routerNgProbeToken},
         config?.initialNavigation ? provideInitialNavigation(config) : [],
-        config?.bindToComponentInputs ? withComponentInputBinding().ɵproviders : [],
```
is this really intended to be deleted?

atscott replied:
Unintentional, thanks for the catch.

atscott commented:
Removing risk label since I ended up going with the sometimes sync, sometimes async behavior so that the default behavior is unchanged.

atscott commented:
[TGP](https://fusion2.corp.google.com/presubmit/555993401/OCL:555993401:BASE:561043376:1693362970318:dfa5ad63/targets) shows no failures for the current approach. I ran a quick test to enable view transitions unconditionally and it appears that test failures are generally due to ZoneJS and the fact that it doesn't patch the `startViewTransition` API so the `fakeAsync` with `flush` and `tick` doesn't work. However, tests can instead use `async/await` along with the `RouterTestingHarness` to resolve these failures.

devversion commented on packages/router/src/navigation_transition.ts:
```diff
@@ -602,37 +603,67 @@ export class NavigationTransitions {
+                             // The above event causes entrance into user code that may error and
+                             // cancel the navigation. Ensure we are still processing before
+                             // continuing.
+                             if (this.currentNavigation?.id !== t.id) {
```
was this then same behavior? (i.e. the cancelling). I assume yes and then the `activateRoutes` map never ran then?

atscott replied:
That's right. The benefit we get from a huge rxjs pipe is that if something causes the transition to be canceled, no further operations will be executed.

atscott replied:
I created an alternate approach that retains the current rxjs operations and instead does some more back-and-forth communication between the navigation transition and the view transition: https://github.com/atscott/angular/commit/70f147d6ba13da60334c2e582888b5f5575bbdc9

atscott committed:
feat(router): Add feature to support the View Transitions API

The View Transitions API enables easy animations when transitioning between different DOM states. This commit adds an opt-in feature to the Router which runs the component activation and deactivation logic in the document.startViewTransition callback. If the browser does not support this API, route activation and deactivation will happen synchronously.

resolves #49401

jessicajaniuk reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented on packages/router/src/utils/view_transition.ts:
```diff
@@ -0,0 +1,60 @@
+import {afterNextRender, InjectionToken, Injector, NgZone} from '@angular/core';
+import {from, Observable, of} from 'rxjs';
+
+export const TRANSITION_HELPER =
```
nit: I'm wondering if we should try to preserve the "view transition" as a term, so that it's easier to understand (without looking at its implementation or docs) what this token refers to.
```suggestion
export const VIEW_TRANSITION_HELPER =
```

AndrewKushnir commented on packages/router/src/navigation_transition.ts:
```diff
@@ -602,37 +603,67 @@ export class NavigationTransitions {
+                           }, this.environmentInjector);
+
+                           // prevent finalize block from running until the code above runs (which
+                           // is async when run inside startViewTransition).
```
The comments that you've added in this function are awesome 👍

AndrewKushnir commented on packages/router/src/navigation_transition.ts:
```diff
@@ -602,37 +603,67 @@ export class NavigationTransitions {
+                             // Activating routes enters user code that could trigger a new
+                             // navigation. Ensure the current navigation is the same as this
+                             // transition before continuing.
+                             if (this.currentNavigation?.id !== t.id) {
```
nit: we can also have a helper function here that would act as an extra inline documentation, something like this:
```suggestion
                             if (!this.wasTransitionInterrupted(t.id)) {
```

atscott replied:
Updated approach to minimize changes in the navigation transition.

AndrewKushnir commented on packages/router/src/provide_router.ts:
```diff
@@ -690,6 +693,34 @@ export function withComponentInputBinding(): ComponentInputBindingFeature {
+ * @returns A set of providers for use with `provideRouter`.
+ * @see https://developer.chrome.com/docs/web-platform/view-transitions/
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API
+ * @experimental
```
I'm not sure this is reflected on AIO, should we use `@developerPreview` instead?

atscott replied:
I'm not sure either but @jelbourn did mention this tag and we believe it's more appropriate given the state of the view transition API in general.

atscott replied:
@jelbourn It does appear that the `@experimental` tag doesn't show up on AIO (https://ng-dev-previews-fw--pr-angular-angular-51314-y5diongs.web.app/api/router/withViewTransitions)

AndrewKushnir commented on packages/router/src/provide_router.ts:
```diff
@@ -690,6 +693,34 @@ export function withComponentInputBinding(): ComponentInputBindingFeature {
 
+/**
+ * Enables view transitions in the Router by running the route activation and deactivation inside of
+ * `document.startViewTransition`.
```
I'd also propose adding a note here to mention that this feature is not available in all browsers yet and Angular would fall back to a different code path if it detects that the feature is missing.

AndrewKushnir reviewed (approved):
Looks great, just a few minor comments.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

atscott committed:
fixup! feat(router): Add feature to support the View Transitions API

AndrewKushnir commented:
This PR was merged into the repository by commit 73e4bf2ed2471faf44a49b591e19a390d5867449.
